pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -637,52 +640,36 @@ func (r *ReconcileMachinePool) syncMachineSets(
 				if rMS.Spec.Template.Spec.Taints == nil {
 					rMS.Spec.Template.Spec.Taints = []corev1.Taint{}
 				}
-				// Make a temporary map of generated MachineSet's taints for easy lookup
-				gTaintsByID := make(map[hivev1.TaintIdentifier]*corev1.Taint)
-				for gIndex, gTaint := range ms.Spec.Template.Spec.Taints {
-					gTaintsByID[identifierForTaint(&gTaint)] = &ms.Spec.Template.Spec.Taints[gIndex]
-				}
-				// Build the list of taints for the remote MachineSet from scratch
+				// First, build a list of remote MachineSet taints that do not have taints needing to be deleted
 				taints := make([]corev1.Taint, 0)
-				// Go through the remote MachineSet's taints, adding/replacing matches from the generated MachineSet
 				for _, rTaint := range rMS.Spec.Template.Spec.Taints {
-					tid := identifierForTaint(&rTaint)
-
-					// If we need to "delete" this one, skip it, but mark modified
-					if taintsToDelete.Has(tid) {
+					if taintsToDelete.Has(identifierForTaint(&rTaint)) {
+						// This entry would be deleted from the final list
 						objectModified = true
 						continue
 					}
-
-					if gTaint, foundTaint := gTaintsByID[tid]; foundTaint && gTaint.Value != rTaint.Value {
-						// Existing taint is modified
-						rTaint = *gTaint
-						objectModified = true
-					}
-
-					// If we get here, one of the following is true:
-					// - The taint was in the generated MachineSet. Whether it was modified or not, it must be included.
-					// - The taint was in the remote MachineSet but not the generated one. However, it wasn't in our
-					//   "owned" list, i.e. it is "remote only". Preserve it.
 					taints = append(taints, rTaint)
-					// We'll want to include any "extras" from the generated MachineSet
-					delete(gTaintsByID, tid)
 				}
-				// Anything remaining from the generated MachineSet is new
-				for _, taint := range gTaintsByID {
-					taints = append(taints, *taint)
-				}
-				sort.SliceStable(taints, func(i, j int) bool {
```

Comment:
Got rid of the sorting simply because I was lazy. Is it needed? Or just good-to-have?

Reply:
Missed this; but addressed it in my comments anyway :)